File: core/java/src/main/java/org/aiddl/core/java/function/type/GenericTypeChecker.java
package org.aiddl.core.java.function.type;

import org.aiddl.core.java.eval.Evaluator;
import org.aiddl.core.java.function.FunctionRegistry;
import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.representation.Substitution;
import org.aiddl.core.java.representation.SymbolicTerm;
import org.aiddl.core.java.representation.Term;

public class GenericTypeChecker implements Function {
	long nextFreeID = 0;
	SymbolicTerm uriBase;
	private Term genTypeDef;
	private Term genArgs;
	private Evaluator eval;
	private FunctionRegistry fReg;

	public GenericTypeChecker(SymbolicTerm uriBase, Term genArgs, Term typeDef, Evaluator eval, FunctionRegistry fReg ) {
		this.uriBase = uriBase;
		this.genArgs = genArgs;
		this.genTypeDef = typeDef;
		this.eval = eval;
		this.fReg = fReg;
	}

	@Override
	public Term apply(Term args) {
		Substitution s = this.genArgs.match(args);
		Term typeDef = genTypeDef.substitute(s);
		SymbolicTerm uri = uriBase.concat(Term.sym(String.valueOf(nextFreeID++)));
		Function typeCheckFun = new TypeChecker(typeDef, this.eval);
		fReg.addFunction(uri, typeCheckFun);
		//return Term.fref(uri, this.fReg);
		return uri;
	}
}

File: core/scala/src/main/scala/org/aiddl/core/scala/function/type/GenericTypeChecker.scala
package org.aiddl.core.scala.function.`type`

import org.aiddl.core.scala.container.Container
import org.aiddl.core.scala.function.{Evaluator, Function, DefaultFunctionUri as D}
import org.aiddl.core.scala.representation.*

protected[scala] class GenericTypeChecker(uri: Sym, pattern: Term, typeDef: Term, e: Evaluator, c: Container) extends Function {
  private  var nextFreeId = 0

  def apply(args: Term): Term = {
    pattern unify args match {
      case Some(s) => {
        this.nextFreeId += 1
        val uri = this.uri + Sym(s"${this.nextFreeId}")
        val fType = new TypeFunction(typeDef\s, e)
        c.addFunction(uri, fType)
        FunRef(uri, fType)
      }
      case None => throw IllegalArgumentException(s"Pattern $args does not match pattern $typeDef of generic type $uri")
    }
  }
}
